### PR TITLE
Adds Alternate Keys vocabulary

### PR DESCRIPTION
**Pull request by christopherabrooks**

### CODE REVIEW

**Comment by markdstafford** [comment]
@ralfhandl, if you're okay with the initial version of this vocabulary, do you mind +1-ing it so I can move the process forward?

File: OData.Community.AlternateKeys.V1.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+  </Term>
+
+  <ComplexType Name="AlternateKey">
+    <Property Type="Collection(Keys.PropertyRef)" Name="Key">
+      <Annotation Term="Core.Description" String="The set of properties that make up this key"/>
+    </Property>
+  </ComplexType>
+
+  <ComplexType Name="PropertyRef">
+    <Property Type="Edm.PropertyPath" Name="Name">
```
Comment (congysu) [inline OData.Community.AlternateKeys.V1.md:58]: Just a side note that: EdmLib for now might not support PropertyPath yet.  Related issue: https://github.com/OData/odata.net/issues/37

File: OData.Community.AlternateKeys.V1.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+Services SHOULD NOT provide an alternate key definition that exactly matches the property set of the primary entity key.
+ 
+###2.2	Client calls a service with alternate keys
+When a client wishes to address an operation to an entity using alternate key values, they will provide the property name(s) and values for the intended key in the key segment.
+
+####2.2.1	Key Predicate Formats and Key Matching
+OData defines 2 types of key predicates, a simpleKey and a compoundKey as follows:
+
+```http
+http://host/service/persons(123) (simpleKey)
```
Comment (congysu) [inline OData.Community.AlternateKeys.V1.md:224]: Any suggestion for 'key as segment' syntax for alternate key?

**Comment by ralfhandl** [comment]
Instead of a "plural" term that is a collection of AlternateKey, one per alternate key, we could make it a "singular" term that can be applied multiple times with different qualifiers. The qualifiers would then be an "identifier" for the alternate key. We could also reduce it to just a collection of property references:

```
<Term AppliesTo="EntityType EntitySet NavigationProperty" Type="Collection(Keys.PropertyRef)" Name="AlternateKey">  
  <Annotation Term="Core.Description" String="An alternate key"/>  
</Term>
```

Example:

```
<EntityType Name="Person">
  <Key>
    <PropertyRef Name="ID" />
  </Key>

  <Annotation Term="Keys.AlternateKey" Qualifier="SSN">
    <Collection>
      <Record>
        <PropertyValue Property="Name" PropertyPath="SSN" />
      </Record>
    </Collection>
  </Annotation>

  <Property Name="ID" Type="Edm.Int64" Nullable="false" />
  <Property Name="SSN" Type="Edm.String" Nullable="true" />
</EntityType>
```

What do you think?

File: OData.Community.AlternateKeys.V1.md (hunk 2 above)
Comment (ralfhandl) [inline OData.Community.AlternateKeys.V1.md:224]: We initially discusses two "key as segment" flavors: name-value segments

```
    http://host/service/Customers/CustomerID=ALFKI
```

and value-only segments:

```
    http://host/service/Customers/ALFKI
```

The latter is the more appealing, but it doesn't allow alternate keys.

**Comment by christopherabrooks** [comment]
We discussed the Qualifier approach and went with the simple collection approach for a few reasons.
1. Keys don't have an "identifier" that has meaning to callers.  They are uniquely identified only by the unordered set of attributes to which they refer.  We felt that requiring implementers to develop a unique public name here added cost and complexity but no value.
2. We felt the meaning of the qualifier / name would be potentially confusing to consumers as relates to how API's should expose the qualifiers / names, or how the "unqualified" item should be treated differently.

**Comment by ralfhandl** [comment]
Good points!
+1 to the current proposal

**Pull request merged**